dra-evolution/pkg/api/claim_types.go

```diff
@@ -269,85 +298,96 @@ type ResourceClaimStatus struct {
 	// +patchMergeKey=uid
 	// +optional
 	ReservedFor []ResourceClaimConsumerReference `json:"reservedFor,omitempty" protobuf:"bytes,3,opt,name=reservedFor" patchStrategy:"merge" patchMergeKey:"uid"`
+
+	// Indicates that a claim is to be deallocated. While this is set,
+	// no new consumers may be added to ReservedFor.
+	//
+	// This is only used if the claim needs to be deallocated by a DRA driver.
+	// That driver then must deallocate this claim and reset the field
+	// together with clearing the Allocation field.
+	//
+	// This is an alpha field and requires enabling the DRAControlPlaneController
+	// feature gate.
+	//
+	// +optional
+	DeallocationRequested bool `json:"deallocationRequested,omitempty" protobuf:"varint,4,opt,name=deallocationRequested"`
 }
 
 // AllocationResult contains attributes of an allocated resource.
 type AllocationResult struct {
-	// DriverData contains the state associated with an allocation that
-	// should be maintained throughout the lifetime of a claim. Each
-	// entry contains data that should be passed to a specific kubelet
-	// plugin once the claim lands on a node.
+	// ControllerName is the name of the DRA driver which handled the
+	// allocation. That driver is also responsible for deallocating the
+	// claim. It is empty when the claim can be deallocated without
+	// involving a driver.
 	//
-	// Setting this field is optional. It has a maximum size of 32 entries.
-	// If empty, nothing was allocated for the claim and kubelet does not
-	// need to prepare anything for it.
+	// A driver may allocate devices provided by other drivers, so this
```

Comment:
That sounds tricky. The two drivers then somehow need to coordinate on the allocations? I don't think we need to get deep into this now, but I'd like to understand what this use case is so if necessary we can eventually fold it into the structured parameters model.

Reply:
The "meta driver" which implements the allocation needs to use the same allocation API as the drivers that it is wrapping. @klueska used this to combine GPUs with NICs. In practice, the "allocation API" was the CRD that those two drivers used to advertise their node-local devices.

I think we have that use case covered in some form with match attributes. The "reserve a device in ResourcePool" could become our standardized "allocation API" for cases where drivers want to support both structured parameters for simple cases and a control plane controller for the more complex scenarios. As it stands now, it's either/or.

Reply:
I think with the partitionable device model, we can handle this completely on the driver side. See https://github.com/kubernetes-sigs/wg-device-management/pull/27#issuecomment-2155167940. Is this still needed in that case?

Reply:
We need the controller name for classic DRA. It cannot be inferred from something else.

Reply:
The point here is that there are two different ResourcePools -- each managed by a "standard" driver for devices of that type. These drivers can be from different vendors and don't have to be aware of each other.

Along side of this, one can then create a "meta" controller which knows how to allocate devices of both types together. So long as it knows how to speak the "language" of each driver in order to perform these allocations atomically, it can sit side by side with them without interfering with the operation of each "standard" driver.

We do not envision such "meta" controllers being for general purpose use, but rather customized to specific environments that need to do non-standard things. This mechanims allows them to install the standard drivers on their clusters, and then only deploy a custom controller to do allocations, rather than having to write an entirely new driver from scratch that also implements the kubelet-plugin side of things.

Reply:
Thanks, I do wonder if the alternative "compound device" design above will cover that use case. Nonetheless, we can keep this here for all classic DRA use cases and we will see if we need it in the long run.